### PR TITLE
Route gemma-4 sliding-window layers through FlashAttention-2

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Gemma-4 mixes 25 sliding-window layers (head_dim 256, window 1024) with 5 global layers (head_dim 512). FlashAttention-2 caps head_dim at 256, so the global layers disable it for the whole model and every layer falls back to SDPA, which spends O(seq^2) compute per sliding layer and dominates the step at long context.

## What this does

A causal sliding window is exactly FA2's `window_size` argument, so intercept the registered sdpa attention function and, for a gemma-4 sliding layer whose head_dim is at most 256, whose dtype is bf16/fp16, and whose mask is the plain causal band (no token padding), call FA2 with `window_size=(w-1, 0)`. Global head_dim 512 layers and padded or non-band masks defer to the original SDPA unchanged. FA2 is an opaque op so the compiled graph is unaffected.

On by default when flash-attn is importable; `UNSLOTH_GEMMA4_FLASH_SLIDING=0` reverts to SDPA.

## Testing

- Result matches SDPA with the explicit band mask to bf16 rounding (forward rel 2.5e-3, backward 3.5e-3); unit test included.
- Measured on a 26B-A4B LoRA SFT (4-bit, gradient checkpointing): 1.66x at 4k, 2.21x at 8k, 2.65x at 16k, 3.11x at 32k, memory unchanged. The speedup grows with sequence length as attention becomes a larger share of the step.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Reworked the mask handling. The plain-band verdict is now stored on the mask tensor instead of an `id()`-keyed dict, since a freed tensor's id can be reused by a new mask and return a stale hit. The band check now verifies the whole mask rather than sampling rows, so a packed-sequence boundary can no longer slip through to FlashAttention. The fast path is only taken when the call is actually causal (an explicit mask is present or `is_causal`), mirroring sdpa, so a bidirectional layer stays bidirectional. Registered the new submodule in the import-smoke list. Tests pass.

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        return False
+    # Cache the verdict on the tensor itself. An id()-keyed dict can alias a
+    # freed mask to a newly allocated one and return a stale result; binding to
+    # the tensor lifetime avoids that. `is not None` preserves a cached False.
+    cached = getattr(mask, "_unsloth_plain_band", None)
+    if cached is not None:
+        return cached
+    if mask.shape[-2] != S or mask.shape[-1] != S:
+        try: mask._unsloth_plain_band = False
+        except (AttributeError, RuntimeError): pass
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: The bare except is intentional. The plain-band verdict is a best-effort cache write onto the mask tensor; a few tensor types reject custom attributes, and a failed write is non-fatal since the next call just recomputes. This matches the existing except: pass usage across temporary_patches.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        return False
+    # Verify the WHOLE mask over every batch element and head. A sampled-row
+    # probe would miss a packed-sequence boundary that falls between the sampled
+    # rows and let FA2 attend across samples. A plain band allows key j for
+    # query i iff i - w < j <= i; anything else (padding, packing) must reject.
+    allowed = mask if mask.dtype == torch.bool else (mask > -1e4)
+    idx = torch.arange(S, device=mask.device)
+    band = (idx[None, :] <= idx[:, None]) & (idx[None, :] > idx[:, None] - w)
+    ok = bool((allowed == band).all().item())
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: Right. The whole-mask check now scans in 1024-row blocks, so the peak transient is O(block x S) instead of a second full S x S band, and it no longer risks OOM at 16k/32k before the SDPA fallback.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/__init__.py [hunk 3]
```diff
@@ -20,6 +20,7 @@
 from .misc import *
 from .gemma3n import *
 from .gemma4 import *
+from .gemma4_flash_sliding import *
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/__init__.py:26]: Added: gemma4_flash_sliding is now imported in the submodule list, so the import-smoke test covers it.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    rows = allowed[probes]                                # (P, S)
+    idx = torch.arange(S, device=dev)[None, :]
+    cnt = rows.sum(-1)
+    minidx = torch.where(rows, idx, torch.full_like(idx, S)).amin(-1)
+    maxidx = torch.where(rows, idx, torch.full_like(idx, -1)).amax(-1)
+    lo = torch.clamp(probes - w + 1, min=0)
+    hi = probes
+    ok = bool(((cnt == (hi - lo + 1)) & (minidx == lo) & (maxidx == hi)).all().item())
+    _BAND_OK[key] = ok
+    return ok
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: Reworked: the plain-band verdict is now stashed on the mask tensor itself, not an id()-keyed dict, so a recycled address can never return a stale hit.

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                              dropout=0.0, scaling=None, is_causal=None, **kwargs):
+    if (_enabled() and _HAS_FA2
+            and getattr(module, "is_sliding", False)
+            and type(module).__name__.startswith("Gemma4")
+            and query.dim() == 4 and query.shape[-1] <= 256
+            and query.dtype in (torch.float16, torch.bfloat16)):
+        w = kwargs.get("sliding_window", None) or getattr(module, "sliding_window", None)
+        Sq = query.shape[2]
+        Sk = key.shape[2]
+        if (w and Sq == Sk and Sq > w and _mask_is_plain_band(attention_mask, Sq, w)):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: Handled. The fast path now requires an explicit mask or a causal call, so a bidirectional layer with no mask stays bidirectional and is not forced causal.

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        return cached
+    if len(_BAND_OK) > 64:
+        _BAND_OK.clear()
+    m = mask[:, 0]
+    if m.shape[-2] != S or m.shape[-1] != S:
+        _BAND_OK[key] = False
+        return False
+    allowed = m if m.dtype == torch.bool else (m > -1e4)
+    dev = m.device
+    cand = sorted({x for x in (0, 1, w - 1, w, w + 1, 2 * w, S // 2, S - 1, S - w, S - w - 1) if 0 <= x < S})
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: The band check now verifies every row of every batch element, not a sampled subset, so a packed or padded boundary can no longer slip through to the fast path.

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+def _mask_is_plain_band(mask, S, w):
+    """True if `mask` is exactly the causal + sliding band with no token padding."""
+    if mask is None:
+        return True
+    if not torch.is_tensor(mask) or mask.dim() != 4:
+        return False
+    key = id(mask)
+    cached = _BAND_OK.get(key)
+    if cached is not None:
+        return cached
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: The verdict is stored on the tensor object, not an id()-keyed dict, so a reused address cannot return a stale verdict for a different mask.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            rows = idx[start : start + _block]                                       # (br,)
+            band = (idx[None, :] <= rows[:, None]) & (idx[None, :] > rows[:, None] - w)  # (br, S)
+            msl = m[:, start : start + _block, :]                                     # (B, br, S)
+            match = (msl == band) if is_bool else torch.where(band, msl == 0, msl <= -1e4)
+            if not bool(match.all().item()):
+                ok = False
+                break
+    try:
+        mask._unsloth_plain_band = ok
+    except Exception:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py:115]: Intentional. The plain-band verdict is a best-effort write onto the mask tensor; a few tensor types reject custom attributes, and a failed write just recomputes next call. Same pattern as the other temporary_patches.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        # would replace it with a boolean mask and silently drop the bias.
+        is_bool = m.dtype == torch.bool
+        idx = torch.arange(S, device=m.device)
+        ok = True
+        for start in range(0, S, _block):
+            rows = idx[start : start + _block]                                       # (br,)
+            band = (idx[None, :] <= rows[:, None]) & (idx[None, :] > rows[:, None] - w)  # (br, S)
+            msl = m[:, start : start + _block, :]                                     # (B, br, S)
+            match = (msl == band) if is_bool else torch.where(band, msl == 0, msl <= -1e4)
+            if not bool(match.all().item()):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py:110]: Fixed. Added a torch.compiler.is_compiling() guard at the top of _mask_is_plain_band so it returns False and defers to the original SDPA while tracing, avoiding the data-dependent .item() graph break and the untraceable tensor-attribute write under fullgraph. Eager behavior is unchanged.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    m = _MASK_CACHE.get(key)
+    if m is not None:
+        return m
+    qi = torch.arange(w, device=device)[:, None]
+    ki = torch.arange(2 * w, device=device)[None, :]
+    base = (ki > qi) & (ki <= qi + w)                     # (w, 2w) local band
+    m = base[None].expand(nb, w, 2 * w).clone()           # (nb, w, 2w)
+    m[0] &= (torch.arange(2 * w, device=device)[None, :] >= w)  # block 0: drop phantom prev-block
+    m = m[:, None]                                         # (nb, 1, w, 2w)
+    _MASK_CACHE[key] = m
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py:59]: Fixed. _block_mask now applies a small FIFO bound (cap 8): before inserting a new key it evicts the oldest entry, so varying sequence lengths can no longer grow _MASK_CACHE without limit. Still a pure recompute-on-miss optimization, so correctness is unchanged.

File: unsloth_zoo/temporary_patches/gemma4_flash_sliding.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                    )
+                    _ENGAGED[0] += 1
+                    if _ENGAGED[0] == 1:
+                        logger.info_once(
+                            f"Unsloth: FlashAttention-2 sliding window engaged for gemma-4 "
+                            f"(window={w})."
+                        )
+                    return out, None                      # (B, S, H, D), no weights
+                except Exception as e:
+                    logger.warning_once(f"Unsloth: gemma-4 FA2 sliding fell back to SDPA ({e})")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_flash_sliding.py]: Fixed. Restructured the router so a runtime FA2 failure now falls through to the banded SDPA kernel before deferring to the original SDPA, instead of skipping the banded fast path. use_fa2 gates the FA2 attempt; on FA2 exception it logs and falls through to a single banded attempt reached by both the non-FA2 and FA2-failure cases, and only a banded failure defers to _ORIG_SDPA.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
